src/txreconciliation.cpp

```diff
@@ -0,0 +1,1048 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <txreconciliation.h>
+
+#include <node/minisketchwrapper.h>
+
+#include <unordered_map>
+#include <util/hasher.h>
+
+namespace {
+
+/** Current protocol version */
+constexpr uint32_t RECON_VERSION = 1;
+/** Static salt component used to compute short txids for sketch construction, see BIP-330. */
+const std::string RECON_STATIC_SALT = "Tx Relay Salting";
+/** Announce transactions via full wtxid to a limited number of inbound and outbound peers. */
+constexpr double INBOUND_FANOUT_DESTINATIONS_FRACTION = 0.1;
+constexpr double OUTBOUND_FANOUT_DESTINATIONS_FRACTION = 0.1;
+/** The size of the field, used to compute sketches to reconcile transactions (see BIP-330). */
+constexpr unsigned int RECON_FIELD_SIZE = 32;
+/**
+ * Allows to infer capacity of a reconciliation sketch based on it's char[] representation,
+ * which is necessary to deserealize a received sketch.
+ */
+constexpr unsigned int BYTES_PER_SKETCH_CAPACITY = RECON_FIELD_SIZE / 8;
+/**
+ * Limit sketch capacity to avoid DoS. This applies only to the original sketches,
+ * and implies that extended sketches could be at most twice the size.
+ */
+constexpr uint32_t MAX_SKETCH_CAPACITY = 2 << 12;
```

Comment:
The value chosen here seems too large based on some tests that i did.

On my laptop (with a decent CPU) a sketch with a capacity of  `2 << 13` and `2 << 13` differences (possible if an extension is requested) took ~20 seconds to decode. On lower end devices this will likely be worse. This is not super bad since sketches are only requested on outbound connections but having the `msghand` thread block like this is undesirable in any case.

I am not entirely sure what an appropriate value would be but given the expected average set difference of 7 i would assume that 256 - 1024 as the max. sketch capacity would suffice.

Reply:
Excellent point. I will address this once I have to rebase this changes.